### PR TITLE
EA-2663 entity stats impmenetation

### PR DESCRIPTION
**Pull request by SrishtiSingh-eu**

### CODE REVIEW

File: entity-stats/src/main/java/eu/europeana/entity/stats/service/UsageStatsService.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+     * @return
+     */
+    private EntityStats getFacetsResults(Query searchQuery) {
+        List<FacetFieldView> facets = solrEntityService.search(searchQuery, null,null, null ).getFacetFields();
+        if (!facets.isEmpty()) {
+            EntityStats entityStats = new EntityStats();
+            // fetch the first facet result view
+            Map<String, Long> map = facets.get(0).getValueCountMap();
+            for (Map.Entry<String, Long> entry : map.entrySet()) {
+                if(entry.getKey().equals("Agent")) {
```
Comment (gsergiu) [inline entity-stats/src/main/java/eu/europeana/entity/stats/service/UsageStatsService.java:106]: we should use EntityTypes.<type>.getInternalTyp() instead for hardcoded values

File: entity-stats/src/main/java/eu/europeana/entity/stats/service/UsageStatsService.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+       entities.setPlaces(getPercentage(entityPerLanguage.getPlaces(), entitystatsTotal.getPlaces()));
+       entities.setConcepts(getPercentage(entityPerLanguage.getConcepts(), entitystatsTotal.getConcepts()));
+       entities.setAgents(getPercentage(entityPerLanguage.getAgents(), entitystatsTotal.getAgents()));
+       entities.setOrganisations(getPercentage(entityPerLanguage.getOrganisations(), entitystatsTotal.getOrganisations()));
+       entities.setTotal(getPercentage(entityPerLanguage.getTotal(), entitystatsTotal.getTotal()));
+    }
+
+    private static float getPercentage(float value, float total) throws UsageStatsException {
+      try {
+          if (total != 0.0) {
```
Comment (gsergiu) [inline entity-stats/src/main/java/eu/europeana/entity/stats/service/UsageStatsService.java:139]: better use > 0, the total should be of type long, should be converted to double only for division reasons if needed (I assume that we don't really need the decimals when computing percentages)

File: entity-web/src/main/java/eu/europeana/entity/web/controller/UsageStatsController.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+
+import javax.servlet.http.HttpServletRequest;
+import java.util.Date;
+
+@Controller
+@Api(tags = "Discovery API")
```
Comment (gsergiu) [inline entity-web/src/main/java/eu/europeana/entity/web/controller/UsageStatsController.java:24]: we should have an own grouping for stats

File: entity-web/src/main/java/eu/europeana/entity/web/jsonld/JsonLdSerializer.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    ObjectMapper mapper;
+    public static final String DATE_FORMAT  = "yyyy-MM-dd'T'HH:mm:ss'Z'";
+
+    public JsonLdSerializer() {
+        mapper = new ObjectMapper();
+        SimpleDateFormat df = new SimpleDateFormat(DATE_FORMAT, Locale.ENGLISH);
+        mapper.setDateFormat(df);
+    }
+
+    public String serializeMetric(EntityMetric metric) throws EntityApiRuntimeException {
```
Comment (gsergiu) [inline entity-web/src/main/java/eu/europeana/entity/web/jsonld/JsonLdSerializer.java:25]: better use a generic implementation public String serialize(Object) ... as the implementation doesn't depend on the type of the object

**Review by gsergiu — COMMENTED** [review]
I left a few comment for improving the implementation

File: entity-stats/src/main/java/eu/europeana/entity/stats/service/UsageStatsService.java (hunk 2 above)
Comment (SrishtiSingh-eu) [reply entity-stats/src/main/java/eu/europeana/entity/stats/service/UsageStatsService.java:139]: Total is indeed not in decimals but as we are using the same model for both percentages values (entity per language) and whole values (entity per type), it's set to float now. it was long before but Hugo suggested having decimal values as well. That gives better insights. Though the format is handled in Databox. But I will use > 0 checks

File: entity-web/src/main/java/eu/europeana/entity/web/controller/UsageStatsController.java (hunk 3 above)
Comment (SrishtiSingh-eu) [reply entity-web/src/main/java/eu/europeana/entity/web/controller/UsageStatsController.java:24]: done - changed to usage Statistics api

File: entity-stats/src/main/java/eu/europeana/entity/stats/service/UsageStatsService.java (hunk 1 above)
Comment (SrishtiSingh-eu) [reply entity-stats/src/main/java/eu/europeana/entity/stats/service/UsageStatsService.java:106]: done

File: entity-web/src/main/java/eu/europeana/entity/web/jsonld/JsonLdSerializer.java (hunk 4 above)
Comment (SrishtiSingh-eu) [reply entity-web/src/main/java/eu/europeana/entity/web/jsonld/JsonLdSerializer.java:25]: done

**Pull request merged**